modules/metadata/mod_remoteip.c

```diff
@@ -571,14 +597,99 @@ static int remoteip_modify_request(request_rec *r)
+                    /* Propagate host header value from the header set by proxy and update virtual host chosen (the primary purpose of it)
+                     */
+                    apr_array_header_t *hdrs = apr_array_make(r->pool, 0, sizeof(char*));
+                    while (*host && (val = ap_get_token(r->pool, &host, 1))) {
+                        *(char **)apr_array_push(hdrs) = apr_pstrdup(r->pool, val);
+                        if (*host != '\0')
+                            ++host;
+                    }
+                    apr_table_setn(r->headers_in, "Host", apr_pstrdup(r->pool, ((char **)hdrs->elts)[((hdrs->nelts)-1)]));
+                    r->hostname = apr_pstrdup(r->pool, ((char **)hdrs->elts)[((hdrs->nelts)-1)]);
+                    ap_update_vhost_from_headers(r);
+                }
+            }
+
+            /* Remote port option, due to the transient nature of subrequests we have to annotate the connection itself for ap_default_port() hook
+             */
+            if (config->port_header_name) {
+                const char *port;
+                if ( (port = apr_table_get(r->headers_in, config->port_header_name) )) {
+                    r->parsed_uri.port = r->server->port;
+                    apr_table_setn(r->connection->notes, "remoteip_port", apr_pstrdup(r->connection->pool, port));
+                } else {
+                    /* Subsequent requests inside a single connection need to have the annotation reset if they do not have the header set
+                     */
+                    apr_table_unset(r->connection->notes, "remoteip_port");
+                }
+            }
+
+            /* Remote HTTPS protocol option, due to transient nature of subrequests we have to annotate the connection itself for ap_conn_is_ssl() hook
+             */
+            if (config->proto_header_name) {
+                const char *proto;
+                if ( (proto = apr_table_get(r->headers_in, config->proto_header_name)) ) {
+                    /* Exact HTTPS protocol ID option not set means just the header presence is enough, otherwise check for a match
+                     */
+                    if (!config->proto_https_enable || !strcmp(proto, config->proto_https_enable)) {
+                        apr_table_setn(r->connection->notes, "remoteip_https", "on");
+                    } else {
+                        /* Subsequent requests inside a single connection need to have the annotation reset if they do not have the content match
+                         */
+                        apr_table_unset(r->connection->notes, "remoteip_https");
+                    }
+                } else {
+                    /* Subsequent requests inside a single connection need to have the annotation reset if they do not have the header set
+                     */
+                    apr_table_unset(r->connection->notes, "remoteip_https");
+                }
+            }
+        }
+    }
```

Comment:
It seems that all the c->notes could go in remoteip_conn_config_t members instead, which would be more efficient.

Reply:
Yes, definitely more efficient indeed, just have to check if it lives through mod_rewrite subrequests mechanics though.